File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.js
function solution(n, costs) {
    // 1. 간선 오름차순 정렬
    costs.sort((a, b) => a[2] - b[2]);

    const parent = Array(n).fill(0).map((_, i) => i);
    
    function find(x){
        if (x === parent[x]) return x;
        return parent[x] = find(parent[x]); // 경로 압축
    };

    function union(a, b){
        const rootA = find(a);
        const rootB = find(b);
        if (rootA === rootB) return false; // 사이클 있음
        parent[rootB] = rootA;
        return true;
    };

    let answer = 0;
    for (const [from, to, cost] of costs) {
        if (union(from, to)) {
            answer += cost; // 간선 채택 시 비용 추가
        }
    }

    return answer;
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;
class Solution {
    static int[] parent;
    public int solution(int n, int[][] costs) {
        Arrays.sort(costs,(a,b)->a[2]-b[2]);
        int answer = 0;
        parent = new int[n];
        for(int i=0; i<n; i++){
            parent[i]=i;
        }
        for(int i=0; i<costs.length; i++){
            if(union(costs[i][0],costs[i][1])){
                answer+=costs[i][2];
            }
        }
        return answer;
    }
    static int find(int x){
        if(x==parent[x]) return x;
        return parent[x]=find(parent[x]);
        
    }
    
    static boolean union(int a, int b){
        int rootA = find(a);
        int rootB = find(b);
        if (rootA == rootB) return false; // 사이클 있음
        parent[rootB] = rootA;
        return true;
    };
    
}